lib/functions/RegularFunctions.ts

```diff
@@ -315,23 +329,23 @@ const STRLEN = {
  */
 const SUBSTR = {
   arity: [ 2, 3 ],
-  overloads: declare()
+  overloads: declare(C.RegularOperator.SUBSTR)
     .onBinaryTyped(
       [ TypeURL.XSD_STRING, TypeURL.XSD_INTEGER ],
-      (source: string, startingLoc: number) => string([ ...source ].slice(startingLoc - 1).join('')),
+      () => (source: string, startingLoc: number) => string([ ...source ].slice(startingLoc - 1).join('')),
     )
     .onBinary(
       [ TypeURL.RDF_LANG_STRING, TypeURL.XSD_INTEGER ],
-      (source: E.LangStringLiteral, startingLoc: E.NumericLiteral) => {
+      () => (source: E.LangStringLiteral, startingLoc: E.NumericLiteral) => {
         const sub = [ ...source.typedValue ].slice(startingLoc.typedValue - 1).join('');
         return langString(sub, source.language);
       },
     )
     .onTernaryTyped([ TypeURL.XSD_STRING, TypeURL.XSD_INTEGER, TypeURL.XSD_INTEGER ],
-      (source: string, startingLoc: number, length: number) =>
+      () => (source: string, startingLoc: number, length: number) =>
         string([ ...source ].slice(startingLoc - 1, length).join('')))
     .onTernary([ TypeURL.RDF_LANG_STRING, TypeURL.XSD_INTEGER, TypeURL.XSD_INTEGER ],
-      (source: E.LangStringLiteral, startingLoc: E.NumericLiteral, length: E.NumericLiteral) => {
+      () => (source: E.LangStringLiteral, startingLoc: E.NumericLiteral, length: E.NumericLiteral) => {
         const sub = [ ...source.typedValue ].slice(startingLoc.typedValue - 1, length.typedValue).join('');
         return langString(sub, source.language);
       })
```

Comment:
Question here before I make an issue about it. I think this implementation is wrong. The third argument of slice is the endIndex. However the third argument of https://www.w3.org/TR/sparql11-query/#func-substr is the length of the substring.
This function is also based on https://www.w3.org/TR/xpath-functions/#func-substring . It talks about arguments being double and sparql talks about integers. Maybe we should provide both implementation? Doesn't hurt anyone?

Reply:
>  I think this implementation is wrong.

That may be wrong indeed. Aren't there tests for this yet?

> It talks about arguments being double and sparql talks about integers. Maybe we should provide both implementation? Doesn't hurt anyone?

I think there is no need to do that. If I understand the XPath spec correctly, doubles are only allowed there for backwards-compatibility, which is a problem we don't really have here.

Reply:
There are tests but they test with arguments someStr 1 1 which is gives the same result.
Alright. Should I edit this in this PR or create an issue handling this small change?

Reply:
I would do it in a separate PR right after this one. (or at least create an issue now)